src/main/java/bio/terra/workspace/service/datareference/utils/DataReferenceValidationUtils.java

```diff
@@ -64,8 +111,54 @@ private void validateSnapshotReference(SnapshotReference ref, AuthenticatedUserR
     }
     if (!dataRepoService.snapshotExists(ref.instanceName(), ref.snapshot(), userReq)) {
       throw new InvalidDataReferenceException(
-          "The given snapshot could not be found in the Data Repo instance provided."
-              + " Verify that your reference was correctly defined and the instance is correct");
+          String.format(
+              "Snapshot %s could not be found in Data Repo instance %s. Verify that your reference was correctly defined and the instance is correct",
+              ref.snapshot(), ref.instanceName()));
+    }
+  }
+
+  /**
+   * Validates that a bucket exists and the user has GET access to it. This makes no other
+   * assumptions about the bucket.
+   */
+  private void validateGoogleBucket(GoogleBucketReference ref, AuthenticatedUserRequest userReq) {
+    DataReferenceValidationUtils.validateBucketName(ref.bucketName());
+    try {
+      // StorageCow.get() returns null if the bucket does not exist or a user does not have access,
+      // which fails validation.
+      BucketCow bucket = crlService.createStorageCow(userReq).get(ref.bucketName());
+      if (bucket == null) {
+        throw new InvalidDataReferenceException(
+            String.format(
+                "Could not access GCS bucket %s. Ensure the name is correct and that you have access.",
+                ref.bucketName()));
+      }
+    } catch (StorageException e) {
+      throw new InvalidDataReferenceException(
+          String.format("Error while trying to access GCS bucket %s", ref.bucketName()), e);
+    }
+  }
```

Comment:
I haven't looked at the innards of CRL - do you know what happens here if the bucket is requester-pays?

Reply:
I haven't seen that case specifically, but CRL is a really thin wrapper around the Google client library. If we're calling GET on a bucket (which I think is a class B operation), it looks like this will fail with a `400 UserProjectMissing` and we'd end up throwing. Is that a usecase we need to support?